heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/MetricsCacheManagerHttpServer.java

```diff
@@ -0,0 +1,148 @@
+  public static final String PATH_EXCEPTIONSUMMARY = "/exceptionsummary";
+  // logger
+  private static final Logger LOG = Logger.getLogger(MetricsCacheManagerHttpServer.class.getName());
+  // http server
+  private HttpServer server = null;
+  // reference to MetricsCache object
+  private MetricsCache metricsCache = null;
+
+  public MetricsCacheManagerHttpServer(MetricsCache cache, int port) throws IOException {
+    metricsCache = cache;
+
+    server = HttpServer.create(new InetSocketAddress(port), 0);
+
+    server.createContext(PATH_STATS, new HandleStatsRequest());
+    server.createContext(PATH_EXCEPTIONS, new HandleExceptionRequest());
+    server.createContext(PATH_EXCEPTIONSUMMARY, new HandleExceptionSummaryRequest());
+  }
+
+  public void start() {
+    server.start();
+  }
+
+  public void stop() {
+    server.stop(0);
+  }
+
+  // compatible with tmaster stat interface: http+protobuf
+  public class HandleStatsRequest implements HttpHandler {
+
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
+      // get the entire stuff
+      byte[] pb = NetworkUtils.readHttpRequestBody(httpExchange);
+      TopologyMaster.MetricRequest req;
+      try {
+        req = TopologyMaster.MetricRequest.parseFrom(pb);
+      } catch (InvalidProtocolBufferException e) {
+        LOG.severe("Unable to decipher data specified in StatsRequest");
+        httpExchange.sendResponseHeaders(400, -1); // throw exception
+        return;
+      }
+      // query cache
+      TopologyMaster.MetricResponse res = metricsCache.GetMetrics(req);
+      // response
+      NetworkUtils.sendHttpResponse(httpExchange, res.toByteArray());
+      // close
+      httpExchange.close();
+      LOG.info("Done with stats request ");
+    }
+  }
+
+  // compatible with tmaster stat interface: http+protobuf
+  public class HandleExceptionRequest implements HttpHandler {
+
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
+      // get the entire stuff
+      byte[] pb = NetworkUtils.readHttpRequestBody(httpExchange);
+      TopologyMaster.ExceptionLogRequest req;
+      try {
+        req = TopologyMaster.ExceptionLogRequest.parseFrom(pb);
+      } catch (InvalidProtocolBufferException e) {
+        LOG.severe("Unable to decipher data specified in ExceptionRequest");
+        httpExchange.sendResponseHeaders(400, -1); // throw exception
+        return;
+      }
+      // query cache
+      TopologyMaster.ExceptionLogResponse res = metricsCache.GetExceptions(req);
+      // response
+      NetworkUtils.sendHttpResponse(httpExchange, res.toByteArray());
+      // close
+      httpExchange.close();
+      LOG.info("Done with exception request ");
+    }
+  }
+
+  // compatible with tmaster stat interface: http+protobuf
+  public class HandleExceptionSummaryRequest implements HttpHandler {
+
+    @Override
+    public void handle(HttpExchange httpExchange) throws IOException {
```

Comment:
These handlers have a lot of copied boilerplate. Could you instead have one generic handler that we reuse that takes an instance of:

```
// T - respuest, U - response
interface RequestHandler<T extends Foo, U extends Foo> {
  T parseRequest(bytes[] requestBytes);
  U generateResponse(T request, MetricsCache metricsCache);
}
```